mirgecom/simutil.py

```diff
@@ -74,92 +88,273 @@ def inviscid_sim_timestep(discr, state, t, dt, cfl, eos,
-    do_status = check_step(step=step, interval=nstatus)
-    if do_viz is False and do_status is False:
-        return 0
+    """Checkpoint the simulation status.
+
+    Checkpoints the simulation status by reporting relevant diagnostic
+    quantities, such as pressure/temperature, and visualization.
+
+    Parameters
+    ----------
+    eos: mirgecom.eos.GasEOS
+        Implementing the pressure and temperature functions for
+        returning pressure and temperature as a function of the state *q*.
+    q
+        State array which expects at least the conserved quantities
+        (mass, energy, momentum) for the fluid at each point. For multi-component
+        fluids, the conserved quantities should include
+        (mass, energy, momentum, species_mass), where *species_mass* is a vector
+        of species masses.
+    freq: nstatus
+        An integer denoting the step frequency for performing status checks.
+    freq: nviz
+        An integer denoting the step frequency for writing vtk output.
+    """
+    exception = None
+    comm = discr.mpi_communicator
+    rank = 0
+    if comm is not None:
+        rank = comm.Get_rank()
+
+    try:
+        # Status checks
+        if check_step(step=step, interval=nstatus):
+            from mirgecom.fluid import split_conserved
+
+            #        if constant_cfl is False:
+            #            current_cfl = get_inviscid_cfl(discr=discr, q=q,
+            #                                           eos=eos, dt=dt)
+
+            cv = split_conserved(discr.dim, q)
+            dependent_vars = eos.dependent_vars(cv)
+            msg = make_status_message(discr=discr,
+                                    t=t, step=step, dt=dt, cfl=cfl,
+                                    dependent_vars=dependent_vars)
+            if rank == 0:
+                logger.info(msg)
+
+            # Check the health of the simulation
+            sim_healthcheck(discr, eos, q, step=step, t=t)
+
+            # Compare with exact solution, if provided
+            if exact_soln is not None:
+                compare_with_analytic_solution(
+                    discr, eos, q, exact_soln,
+                    step=step, t=t, exittol=exittol
+                )
+
+        # Visualization
+        if check_step(step=step, interval=nviz):
+            sim_visualization(
+                discr, eos, q, visualizer, vizname,
+                step=step, t=t,
+                exact_soln=exact_soln, viz_fields=viz_fields,
+                overwrite=overwrite, vis_timer=vis_timer
+            )
+    except SynchronizedError as err:
+        exception = err
+
+    terminate = True if exception is not None else False
+
+    if comm is None:
+        if terminate:
+            raise exception
+        return
+
+    from mpi4py import MPI
+
+    terminate = comm.allreduce(terminate, MPI.LOR)
+
+    if terminate:
+        # Log crash error message
+        if rank == 0:
+            logger.info(str(exception))
```

Comment:
I think this will need to be tweaked, because `exception` doesn't necessarily exist on rank 0.